packages/assets-controllers/src/AssetsContractController.ts

```diff
@@ -62,137 +67,254 @@ export const SINGLE_CALL_BALANCES_ADDRESS_BY_CHAINID: Record<Hex, string> = {
     '0x6aa75276052d96696134252587894ef5ffa520af',
   [SupportedTokenDetectionNetworks.moonriver]:
     '0x6aa75276052d96696134252587894ef5ffa520af',
-};
+} as const satisfies Record<Hex, string>;
 
 export const MISSING_PROVIDER_ERROR =
   'AssetsContractController failed to set the provider correctly. A provider must be set for this method to be available';
 
-/**
- * @type AssetsContractConfig
- *
- * Assets Contract controller configuration
- * @property provider - Provider used to create a new web3 instance
- */
-// This interface was created before this ESLint rule was added.
-// Convert to a `type` in a future major version.
-// eslint-disable-next-line @typescript-eslint/consistent-type-definitions
-export interface AssetsContractConfig extends BaseConfig {
-  provider: Provider | undefined;
-  ipfsGateway: string;
-  chainId: Hex;
-}
-
 /**
  * @type BalanceMap
  *
  * Key value object containing the balance for each tokenAddress
  * @property [tokenAddress] - Address of the token
  */
-// This interface was created before this ESLint rule was added.
-// Convert to a `type` in a future major version.
-// eslint-disable-next-line @typescript-eslint/consistent-type-definitions
-export interface BalanceMap {
+export type BalanceMap = {
   [tokenAddress: string]: BN;
-}
+};
+
+/**
+ * The name of the {@link AssetsContractController}
+ */
+const name = 'AssetsContractController';
+
+/**
+ * A utility type that derives the public method names of a given messenger consumer class,
+ * and uses it to generate the class's internal messenger action types.
+ * @template Controller - A messenger consumer class.
+ */
+// TODO: Figure out generic constraint and move to base-controller
+type ControllerActionsMap<Controller> = {
+  [ClassMethod in keyof Controller as Controller[ClassMethod] extends ActionConstraint['handler']
+    ? ClassMethod
+    : never]: {
+    type: `${typeof name}:${ClassMethod & string}`;
+    handler: Controller[ClassMethod];
+  };
+};
+
+type AssetsContractControllerActionsMap =
+  ControllerActionsMap<AssetsContractController>;
+
+/**
+ * The union of all public class method names of {@link AssetsContractController}.
+ */
+type AssetsContractControllerMethodName =
+  keyof AssetsContractControllerActionsMap;
+
+/**
+ * The union of all internal messenger actions available to the {@link AssetsContractControllerMessenger}.
+ */
+export type AssetsContractControllerActions =
+  AssetsContractControllerActionsMap[AssetsContractControllerMethodName];
```

Comment:
Proof of concept for new base-controller utility type that should be useful when creating new controllers, or adding a large number of new actions/methods.